File: src/widgets/fields/coreWidgetFields.ts
import type { ModuleType,
  BudgetData,
  ChecklistData,
  CountdownData,
  CounterData,
  GoalTrackerData,
  HabitData,
  LinkedListData,
  NotesData,
  PollData,
  ProgressData,
  QuoteData,
  RatingData,
  StickyNoteData,
  StopwatchData,
  TimerData,
  TimekeeperData,
  AtlasWidgetData,
} from '../../types/spatial'
import type { FieldDescriptor } from '../contracts/fields'
import { num, text, bool, primaryZoneTime } from './valueHelpers'
import {
  linkedListNodes,
  linkedNodeIndex,
  selectedLinkedNodeId,
  writeCurrentLinkedValue,
} from '../../components/widgets/modules/linkedListSkinModel'

/** Everyday widget fields (notes … tracker). Extracted verbatim from fields.ts; field order IS port-slot order — never reorder within an entry. */
export const CORE_WIDGET_FIELDS = {
  notes: [
    {
      key: 'text',
      label: 'Text',
      valueType: 'text',
      get: (d) => (d as NotesData).text,
      set: (d, v) => ({ ...(d as NotesData), text: text(v) }),
    },
  ],
  linked_list: [
    {
      key: 'head',
      label: 'Head value',
      valueType: 'text',
      get: (d) => linkedListNodes((d as LinkedListData).nodes)[0]?.value ?? '',
    },
    {
      key: 'tail',
      label: 'Tail value',
      valueType: 'text',
      get: (d) => linkedListNodes((d as LinkedListData).nodes).at(-1)?.value ?? '',
    },
    {
      key: 'current',
      label: 'Current value',
      valueType: 'text',
      get: (d) => {
        const data = d as LinkedListData
        const nodes = linkedListNodes(data.nodes)
        return nodes[linkedNodeIndex(nodes, selectedLinkedNodeId(nodes, data.selectedId))]?.value ?? ''
      },
      set: (d, value) => writeCurrentLinkedValue(d as LinkedListData, text(value)),
    },
    {
      key: 'length',
      label: 'Node count',
      valueType: 'number',
      unit: 'count',
      get: (d) => linkedListNodes((d as LinkedListData).nodes).length,
    },
    {
      key: 'empty',
      label: 'Is empty',
      valueType: 'boolean',
      get: (d) => linkedListNodes((d as LinkedListData).nodes).length === 0,
    },
  ],
  sticky_note: [
    {
      key: 'text',
      label: 'Text',
      valueType: 'text',
      get: (d) => (d as StickyNoteData).text,
      set: (d, v) => ({ ...(d as StickyNoteData), text: text(v) }),
    },
  ],
  quote: [
    {
      key: 'text',
      label: 'Quote',
      valueType: 'text',
      get: (d) => (d as QuoteData).text,
      set: (d, v) => ({ ...(d as QuoteData), text: text(v) }),
    },
  ],
  counter: [
    {
      key: 'count',
      label: 'Count',
      valueType: 'number',
      unit: 'count',
      get: (d) => (d as CounterData).count,
      set: (d, v) => ({ ...(d as CounterData), count: num(v) }),
    },
  ],
  progress: [
    {
      key: 'percent',
      label: 'Percent',
      valueType: 'number',
      unit: 'percent',
      get: (d) => (d as ProgressData).percent,
      set: (d, v) => ({
        ...(d as ProgressData),
        percent: Math.min(100, Math.max(0, Math.round(num(v)))),
      }),
    },
  ],
  rating: [
    {
      key: 'value',
      label: 'Stars',
      valueType: 'number',
      get: (d) => (d as RatingData).value,
      set: (d, v) => ({ ...(d as RatingData), value: Math.min(5, Math.max(0, Math.round(num(v)))) }),
    },
  ],
  budget: [
    {
      key: 'total',
      label: 'Total',
      valueType: 'number',
      unit: 'currency',
      get: (d) => (d as BudgetData).items.reduce((s, i) => s + (Number.isFinite(i.amount) ? i.amount : 0), 0),
    },
  ],
  checklist: [
    {
      key: 'done_count',
      label: 'Done count',
      valueType: 'number',
      unit: 'count',
      get: (d) => (d as ChecklistData).items.filter((i) => i.done).length,
    },
    {
      key: 'all_done',
      label: 'All done',
      valueType: 'boolean',
      get: (d) => {
        const items = (d as ChecklistData).items
        return items.length > 0 && items.every((i) => i.done)
      },
    },
  ],
  goal_tracker: [
    {
      key: 'percent',
      label: 'Progress %',
      valueType: 'number',
      unit: 'percent',
      get: (d) => {
        const goal = d as GoalTrackerData
        if (goal.mode === 'simple') return goal.simple?.percent ?? 0
        if (goal.mode === 'hours') return goal.hours && goal.hours.targetHours > 0
          ? Math.min(100, Math.round(goal.hours.loggedHours / goal.hours.targetHours * 100))
          : 0
        if (goal.mode === 'okr') {
          const results = goal.okr?.keyResults ?? []
          const weight = results.reduce((sum, item) => sum + Math.max(0, item.weight), 0) || 1
          return results.reduce((sum, item) => sum + Math.min(1, Math.max(0, item.current / Math.max(1, item.target))) * Math.max(0, item.weight), 0) / weight * 100
        }
        const ms = goal.milestones
        return ms.length === 0 ? 0 : Math.round((ms.filter((m) => m.done).length / ms.length) * 100)
      },
      set: (d, v) => {
        const goal = d as GoalTrackerData
        return {
          ...goal,
          simple: {
            label: (goal.simple?.label ?? goal.goal) || 'Progress',
            percent: Math.min(100, Math.max(0, Math.round(num(v)))),
          },
        }
      },
    },
    {
      key: 'complete',
      label: 'Complete',
      valueType: 'boolean',
      get: (d) => {
        const goal = d as GoalTrackerData
        if (goal.mode === 'simple') return (goal.simple?.percent ?? 0) >= 100
        if (goal.mode === 'hours') return Boolean(goal.hours && goal.hours.targetHours > 0 && goal.hours.loggedHours >= goal.hours.targetHours)
        if (goal.mode === 'okr') return Boolean(goal.okr?.keyResults.length && goal.okr.keyResults.every((item) => item.current >= item.target))
        return goal.milestones.length > 0 && goal.milestones.every((m) => m.done)
      },
    },
  ],
  poll: [
    {
      key: 'votes',
      label: 'Total votes',
      valueType: 'number',
      unit: 'count',
      get: (d) => (d as PollData).options.reduce((s, o) => s + o.votes, 0),
    },
  ],
  habit: [
    {
      key: 'streak',
      label: 'Days done',
      valueType: 'number',
      unit: 'count',
      get: (d) => (d as HabitData).days.filter(Boolean).length,
    },
  ],
  countdown: [
    {
      key: 'days_left',
      label: 'Days left',
      valueType: 'number',
      unit: 'count',
      get: (d) =>
        Math.ceil(
          (new Date((d as CountdownData).targetDate).getTime() - Date.now()) / 86_400_000,
        ) || 0,
      timeSensitive: true,
    },
    {
      key: 'days_until',
      label: 'Days until',
      valueType: 'number',
      unit: 'count',
      get: (d) =>
        Math.ceil(
          (new Date((d as CountdownData).targetDate).getTime() - Date.now()) / 86_400_000,
        ) || 0,
      timeSensitive: true,
    },
  ],
  stopwatch: [
    {
      key: 'running',
      label: 'Running',
      valueType: 'boolean',
      get: (d) => (d as StopwatchData).startedAt !== null,
    },
  ],
  timer: [
    {
      key: 'running',
      label: 'Running',
      valueType: 'boolean',
      get: (d) => (d as TimerData).endAt !== null,
    },
    { key:'mode',label:'Mode',valueType:'text',get:()=> 'countdown' },
  ],
  timekeeper: [
    {
      key: 'running',
      label: 'Running',
      valueType: 'boolean',
      get: (d) => {
        const value=d as TimekeeperData
        if(value.mode==='pomodoro')return value.pomodoro.endAt!==null
        if(value.mode==='stopwatch')return value.stopwatch.startedAt!==null
        if(value.mode==='lap_timer')return value.stopwatch.startedAt!==null
        if(value.mode==='intervals'||value.mode==='tabata'||value.mode==='multi_stage_timer') {
          return typeof value.skinStates?.[value.mode]?.endAt === 'number'
        }
        if(value.mode==='deadline'||value.mode==='world_clock'||value.mode==='chess_clock')return false
        return value.countdown.endAt!==null
      },
    },
    { key:'mode',label:'Mode',valueType:'text',get:(d)=>(d as TimekeeperData).mode },
    {
      key: 'days_left',
      label: 'Days left',
      valueType: 'number',
      unit: 'count',
      timeSensitive: true,
      get: (d) => {
        const target = (d as TimekeeperData).deadline?.targetDate
        if (!target) return 0
        return Math.ceil((new Date(`${target}T00:00:00`).getTime() - Date.now()) / 86_400_000) || 0
      },
    },
    {
      key: 'days_until',
      label: 'Days until',
      valueType: 'number',
      unit: 'count',
      timeSensitive: true,
      get: (d) => {
        const target = (d as TimekeeperData).deadline?.targetDate
        if (!target) return 0
        return Math.ceil((new Date(`${target}T00:00:00`).getTime() - Date.now()) / 86_400_000) || 0
      },
    },
    {
      key: 'sessions_done',
      label: 'Sessions done',
      valueType: 'number',
      unit: 'count',
      get: (d) => (d as TimekeeperData).pomodoro.completed,
    },
    {
      key: 'completed',
      label: 'Completed sessions',
      valueType: 'number',
      unit: 'count',
      get: (d) => (d as TimekeeperData).pomodoro.completed,
    },
    {
      key: 'primary_time',
      label: 'Primary time',
      valueType: 'text',
      timeSensitive: true,
      get: (d) => primaryZoneTime((d as TimekeeperData).worldClock?.zones ?? []),
    },
    {
      key: 'zone_count',
      label: 'Time zones',
      valueType: 'number',
      unit: 'count',
      get: (d) => (d as TimekeeperData).worldClock?.zones.length ?? 0,
    },
  ],
  tracker: [
    { key:'current',label:'Current',valueType:'number',get:(d)=>(d as AtlasWidgetData).primary,set:(d,v)=>({...d as AtlasWidgetData,primary:num(v)}) },
    { key:'target',label:'Target',valueType:'number',get:(d)=>(d as AtlasWidgetData).target,set:(d,v)=>({...d as AtlasWidgetData,target:num(v)}) },
    { key:'active',label:'Active',valueType:'boolean',get:(d)=>(d as AtlasWidgetData).enabled,set:(d,v)=>({...d as AtlasWidgetData,enabled:bool(v)}) },
  ],
} satisfies Partial<Record<ModuleType, FieldDescriptor[]>>

File: apple/Sources/GrovepadCore/Fields/CoreWidgetFields.swift
import Foundation

// ---------------------------------------------------------------------------
// Everyday widget fields (`widgets/fields/coreWidgetFields.ts`): text,
// counter, rating, checklist, goal_tracker, habit, timekeeper. One entry per
// field in the TypeScript order — field order IS port-slot order.
// ---------------------------------------------------------------------------

enum CoreWidgetFields {
    static let tables: [String: [FieldDescriptor]] = [
        "text": [
            FieldDescriptor(
                key: "text", label: "Text", valueType: .text,
                get: { $0.fieldValue("text", default: .text("")) },
                set: { data, value, _ in data.assigning("text", .string(text(value))) }
            ),
        ],
        "counter": [
            FieldDescriptor(
                key: "count", label: "Count", valueType: .number, unit: .count,
                get: { $0.fieldValue("count", default: .number(0)) },
                set: { data, value, _ in data.assigning("count", .number(num(value))) }
            ),
        ],
        "rating": [
            FieldDescriptor(
                key: "value", label: "Rating", valueType: .number,
                get: { $0.fieldValue("value", default: .number(0)) },
                set: { data, value, _ in data.assigning("value", .number(Swift.min(5, Swift.max(0, jsRound(num(value)))))) }
            ),
        ],
        "checklist": [
            FieldDescriptor(
                key: "done_count", label: "Done count", valueType: .number, unit: .count,
                get: { data in .number(Double(data.records("items").filter { JavaScript.truthy($0["done"]) }.count)) }
            ),
            FieldDescriptor(
                key: "all_done", label: "All done", valueType: .boolean,
                get: { data in
                    let items = data.records("items")
                    return .bool(!items.isEmpty && items.allSatisfy { JavaScript.truthy($0["done"]) })
                }
            ),
        ],
        "goal_tracker": [
            FieldDescriptor(
                key: "percent", label: "Progress %", valueType: .number, unit: .percent,
                get: { goal in .number(goalPercent(goal)) },
                set: { goal, value, _ in
                    var simple = JSONObject()
                    simple["label"] = goalSimpleLabel(goal)
                    simple["percent"] = .number(Swift.min(100, Swift.max(0, jsRound(num(value)))))
                    return goal.assigning("simple", .object(simple))
                }
            ),
            FieldDescriptor(
                key: "complete", label: "Complete", valueType: .boolean,
                get: { goal in .bool(goalComplete(goal)) }
            ),
        ],
        "habit": [
            FieldDescriptor(
                key: "streak", label: "Days done", valueType: .number, unit: .count,
                get: { data in .number(Double(data.items("days").filter { JavaScript.truthy($0) }.count)) }
            ),
        ],
        "timekeeper": [
            FieldDescriptor(
                key: "running", label: "Running", valueType: .boolean,
                get: { data in .bool(timekeeperRunning(data)) }
            ),
            FieldDescriptor(
                key: "mode", label: "Mode", valueType: .text,
                get: { $0.fieldValue("mode", default: .text("")) }
            ),
            FieldDescriptor(
                key: "days_left", label: "Days left", valueType: .number, unit: .count, timeSensitive: true,
                get: { data in .number(deadlineDaysLeft(data)) }
            ),
            FieldDescriptor(
                key: "days_until", label: "Days until", valueType: .number, unit: .count, timeSensitive: true,
                get: { data in .number(deadlineDaysLeft(data)) }
            ),
            FieldDescriptor(
                key: "sessions_done", label: "Sessions done", valueType: .number, unit: .count,
                get: { data in (data.object("pomodoro") ?? JSONObject()).fieldValue("completed", default: .number(0)) }
            ),
            FieldDescriptor(
                key: "completed", label: "Completed sessions", valueType: .number, unit: .count,
                get: { data in (data.object("pomodoro") ?? JSONObject()).fieldValue("completed", default: .number(0)) }
            ),
            FieldDescriptor(
                key: "primary_time", label: "Primary time", valueType: .text, timeSensitive: true,
                get: { data in .text(primaryZoneTime(worldClockZones(data))) }
            ),
            FieldDescriptor(
                key: "zone_count", label: "Time zones", valueType: .number, unit: .count,
                get: { data in .number(Double((data.object("worldClock")?.array("zones") ?? []).count)) }
            ),
        ],
    ]

    // MARK: goal_tracker

    static func goalPercent(_ goal: JSONObject) -> Double {
        let mode = goal.string("mode")
        if mode == "simple" { return goal.object("simple")?.number("percent") ?? 0 }
        if mode == "hours" {
            guard let hours = goal.object("hours"), hours.dbl("targetHours") > 0 else { return 0 }
            return Swift.min(100, jsRound(hours.dbl("loggedHours") / hours.dbl("targetHours") * 100))
        }
        if mode == "okr" {
            let results = goal.object("okr")?.records("keyResults") ?? []
            var weight = results.reduce(0.0) { $0 + Swift.max(0, $1.dbl("weight")) }
            if weight == 0 { weight = 1 } // `|| 1`
            let sum = results.reduce(0.0) { sum, item in
                sum + Swift.min(1, Swift.max(0, item.dbl("current") / Swift.max(1, item.dbl("target")))) * Swift.max(0, item.dbl("weight"))
            }
            return sum / weight * 100
        }
        let milestones = goal.records("milestones")
        if milestones.isEmpty { return 0 }
        let done = milestones.filter { JavaScript.truthy($0["done"]) }.count
        return jsRound(Double(done) / Double(milestones.count) * 100)
    }

    static func goalComplete(_ goal: JSONObject) -> Bool {
        let mode = goal.string("mode")
        if mode == "simple" { return (goal.object("simple")?.number("percent") ?? 0) >= 100 }
        if mode == "hours" {
            guard let hours = goal.object("hours") else { return false }
            return hours.dbl("targetHours") > 0 && hours.dbl("loggedHours") >= hours.dbl("targetHours")
        }
        if mode == "okr" {
            let results = goal.object("okr")?.records("keyResults") ?? []
            return !results.isEmpty && results.allSatisfy { $0.dbl("current") >= $0.dbl("target") }
        }
        let milestones = goal.records("milestones")
        return !milestones.isEmpty && milestones.allSatisfy { JavaScript.truthy($0["done"]) }
    }

    /// `(goal.simple?.label ?? goal.goal) || 'Progress'` — `??` falls through
    /// null and missing, `||` falls through every falsy value.
    static func goalSimpleLabel(_ goal: JSONObject) -> JSONValue {
        let label = JavaScript.coalesce(goal.object("simple")?["label"], goal["goal"])
        return JavaScript.truthy(label) ? label! : .string("Progress")
    }

    // MARK: timekeeper

    static func timekeeperRunning(_ data: JSONObject) -> Bool {
        let mode = data.string("mode") ?? ""
        // `x.endAt !== null`: a missing slot is `undefined`, which is not null.
        func notNull(_ object: JSONObject?, _ key: String) -> Bool {
            guard let object else { return false } // the web would throw here
            return object[key].map { !$0.isNull } ?? true
        }
        switch mode {
        case "pomodoro": return notNull(data.object("pomodoro"), "endAt")
        case "stopwatch", "lap_timer": return notNull(data.object("stopwatch"), "startedAt")
        case "intervals", "tabata", "multi_stage_timer":
            return data.object("skinStates")?.object(mode)?.number("endAt") != nil
        case "deadline", "world_clock", "chess_clock": return false
        default: return notNull(data.object("countdown"), "endAt")
        }
    }

    static func worldClockZones(_ data: JSONObject) -> [String] {
        (data.object("worldClock")?.array("zones") ?? []).compactMap(\.stringValue)
    }

    /// `Math.ceil((new Date(`${target}T00:00:00`).getTime() - Date.now()) / 86_400_000) || 0`.
    /// A descriptor getter has no clock parameter, so this reads the system
    /// clock and the current calendar/time zone, exactly as the web reads
    /// `Date.now()` and the browser's local zone. Unparseable targets are 0.
    static func deadlineDaysLeft(_ data: JSONObject) -> Double {
        guard let target = data.object("deadline")?.string("targetDate"), !target.isEmpty else { return 0 }
        guard let midnight = localMidnight(ofDayKey: target) else { return 0 }
        let millis = (midnight.timeIntervalSince1970 * 1000).rounded(.down) - (Date().timeIntervalSince1970 * 1000).rounded(.down)
        let days = (millis / 86_400_000).rounded(.up)
        return days.isFinite && days != 0 ? days : 0
    }

    /// `new Date("YYYY-MM-DDT00:00:00")` — a date-time form without an offset
    /// is local time.
    static func localMidnight(ofDayKey key: String) -> Date? {
        let parts = key.split(separator: "-", omittingEmptySubsequences: false)
        guard parts.count == 3, let year = Int(parts[0]), let month = Int(parts[1]), let day = Int(parts[2]) else { return nil }
        var components = DateComponents()
        components.year = year
        components.month = month
        components.day = day
        components.hour = 0
        components.minute = 0
        components.second = 0
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone.current
        guard calendar.date(from: components) != nil, month >= 1, month <= 12, day >= 1, day <= 31 else { return nil }
        return calendar.date(from: components)
    }

    /// `primaryZoneTime`: the first zone's clock as `en-GB` HH:mm, or `--:--`.
    static func primaryZoneTime(_ zones: [String]) -> String {
        guard let zone = zones.first, let timeZone = TimeZone(identifier: zone) else { return "--:--" }
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_GB")
        formatter.timeZone = timeZone
        formatter.dateFormat = "HH:mm"
        return formatter.string(from: Date())
    }

    /// `isValidTimeZone`: whether the platform knows the identifier.
    static func isValidTimeZone(_ zone: String) -> Bool {
        TimeZone(identifier: zone) != nil
    }
}
